File: app/src/main/java/jlox/Scanner.java
package jlox;

import static jlox.TokenType.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Converts source code into tokens
 */
class Scanner {

    // Original source code
    private final String source;

    // Processed tokens
    private final List<Token> tokens = new ArrayList<>();

    // Variables to track where scanner is in source
    private int start = 0;
    private int current = 0;
    private int line = 1;

    // Reserved language keywords
    private static final Map<String, TokenType> keywords;

    static {
        keywords = new HashMap<>();
        keywords.put("and", AND);
        keywords.put("class", CLASS);
        keywords.put("else", ELSE);
        keywords.put("false", FALSE);
        keywords.put("for", FOR);
        keywords.put("fun", FUN);
        keywords.put("if", IF);
        keywords.put("nil", NIL);
        keywords.put("or", OR);
        keywords.put("print", PRINT);
        keywords.put("return", RETURN);
        keywords.put("super", SUPER);
        keywords.put("this", THIS);
        keywords.put("true", TRUE);
        keywords.put("var", VAR);
        keywords.put("while", WHILE);
    }

    Scanner(String source) {
        this.source = source;
    }

    /**
     * Scans tokens of source until end
     * 
     * @return processed tokens
     */
    List<Token> scanTokens() {
        while (!isAtEnd()) {
            start = current;
            scanToken();
        }

        tokens.add(new Token(EOF, "", null, line));
        return tokens;
    }

    private void scanToken() {
        char c = advance();

        switch (c) {
            case '(':
                addToken(LEFT_PAREN);
                break;
            case ')':
                addToken(RIGHT_PAREN);
                break;
            case '{':
                addToken(LEFT_BRACE);
                break;
            case '}':
                addToken(RIGHT_BRACE);
                break;
            case ',':
                addToken(COMMA);
                break;
            case '.':
                addToken(DOT);
                break;
            case '-':
                addToken(MINUS);
                break;
            case '+':
                addToken(PLUS);
                break;
            case ';':
                addToken(SEMICOLON);
                break;
            case '*':
                addToken(STAR);
                break;

            // Multiple character lexemes
            case '!':
                addToken(match('=') ? BANG_EQUAL : BANG);
                break;
            case '=':
                addToken(match('=') ? EQUAL_EQUAL : EQUAL);
                break;
            case '<':
                addToken(match('=') ? LESS_EQUAL : LESS);
                break;
            case '>':
                addToken(match('=') ? GREATER_EQUAL : GREATER);
                break;
            case '/':
                if (match('/')) {
                    // A comment goes until end of line
                    while (peek() != '\n' && !isAtEnd()) {
                        advance();
                    }
                } else {
                    addToken(SLASH);
                }
                break;
            case ' ':
            case '\r':
            case '\t':
                // Ignore whitespace.
                break;

            case '\n':
                line++;
                break;

            // Literals
            case '"':
                string();
                break;

            default:
                if (isDigit(c)) {
                    number();
                } else if (isAlpha(c)) {
                    identifier();
                } else {
                    Lox.error(line, "Unexpected character.");
                }
                break;
        }
    }

    /**
     * Scans identifiers
     */
    private void identifier() {
        while (isAlphaNumeric(peek()))
            advance();

        String text = source.substring(start, current);
        TokenType type = keywords.get(text);
        if (type == null) type = IDENTIFIER;
        addToken(type);
    }

    /**
     * Creates a number literal
     */
    private void number() {
        while (isDigit(peek()))
            advance();

        // Looking for fractional
        if (peek() == '.' && isDigit(peekNext())) {
            // Consume '.'
            advance();

            while (isDigit(peek()))
                advance();
        }

        addToken(NUMBER, Double.parseDouble(source.substring(start, current)));
    }

    /**
     * Creates a string literal
     */
    private void string() {
        while (peek() != '"' && !isAtEnd()) {
            if (peek() == '\n')
                line++;
            advance();
        }

        if (isAtEnd()) {
            Lox.error(line, "Unterminated string.");
            return;
        }

        // Closing "
        advance();

        // Trim quotes
        String value = source.substring(start + 1, current - 1);
        addToken(STRING, value);
    }

    /**
     * Match an expected character
     * 
     * @param expected
     * @return if chracter is the expected character
     */
    private boolean match(char expected) {
        if (isAtEnd())
            return false;
        if (source.charAt(current) != expected)
            return false;

        current++;
        return true;
    }

    /**
     * Peek one character ahead
     * 
     * @return character
     */
    private char peek() {
        if (isAtEnd())
            return '\0';
        return source.charAt(current);
    }

    /**
     * Peek two character ahead
     * 
     * @return character
     */
    private char peekNext() {
        if (current + 1 >= source.length())
            return '\0';
        return source.charAt(current + 1);
    }

    /**
     * Tests if a char is a letter
     * 
     * @param c
     * @return
     */
    private boolean isAlpha(char c) {
        return (c >= 'a' && c <= 'z') ||
                (c >= 'A' && c <= 'Z') ||
                c == '_';
    }

    /**
     * Tests if a char is a digit or letter
     * 
     * @param c
     * @return
     */
    private boolean isAlphaNumeric(char c) {
        return isAlpha(c) || isDigit(c);
    }

    /**
     * Tests if a char is a digit
     * 
     * @param c
     * @return
     */
    private boolean isDigit(char c) {
        return c >= '0' && c <= '9';
    }

    /**
     * Computes if we have reached end of source
     * 
     * @return if at end of source
     */
    private boolean isAtEnd() {
        return current >= source.length();
    }

    /**
     * Advance one character in the source
     * 
     * @return next character
     */
    private char advance() {
        return source.charAt(current++);
    }

    /**
     * Adds a token to tokens
     * 
     * @param type of token to add
     */
    private void addToken(TokenType type) {
        addToken(type, null);
    }

    /**
     * Adds a token to tokens with an attached literal
     * 
     * @param type    of token to add
     * @param literal value of the token
     */
    private void addToken(TokenType type, Object literal) {
        String text = source.substring(start, current);
        tokens.add(new Token(type, text, literal, line));
    }
}

File: rust/src/scanner.rs
use std::collections::HashMap;

/// Represents a token in the source code.
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub enum TokenType {
    // Single-character tokens.
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    Comma,
    Dot,
    Minus,
    Plus,
    Semicolon,
    Slash,
    Star,

    // One or two character tokens.
    Bang,
    BangEqual,
    Equal,
    EqualEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,

    // Literals.
    Identifier,
    String,
    Number,

    // Keywords.
    And,
    Class,
    Else,
    False,
    For,
    Fun,
    If,
    Nil,
    Or,
    Print,
    Return,
    Super,
    This,
    True,
    Var,
    While,

    Error,
    Eof,
}

/// Represents a token in the source code.
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub struct Token<'source> {
    /// The type of the token.
    pub kind: TokenType,
    /// The lexeme of the token.
    pub lexeme: &'source str,
    /// The line number where the token starts.
    pub line: usize,
    /// Span of the token (start and end on line)
    pub span: (usize, usize),
}

impl<'source> Token<'source> {
    fn make_token(
        kind: TokenType,
        lexeme: &'source str,
        line: usize,
        start: usize,
        end: usize,
    ) -> Self {
        Token {
            kind,
            lexeme,
            line,
            span: (start, end),
        }
    }

    pub fn is_token(&self, kind: TokenType) -> bool {
        self.kind == kind
    }
}

/// Represents a scanner to turn source code into tokens.
pub struct Scanner<'source> {
    /// The source code being scanned.
    pub source: &'source str,
    /// The starting index of the current token.
    start: usize,
    /// The current index in the source code.
    current: usize,
    /// The current line number.
    line: usize,

    /// The keywords of the language
    keywords: HashMap<&'source str, TokenType>,
}

impl<'source> Scanner<'source> {
    pub fn new(source: &'source str) -> Self {
        Scanner {
            source,
            start: 0,
            current: 0,
            line: 1,
            keywords: HashMap::from([
                ("and", TokenType::And),
                ("class", TokenType::Class),
                ("else", TokenType::Else),
                ("false", TokenType::False),
                ("for", TokenType::For),
                ("fun", TokenType::Fun),
                ("if", TokenType::If),
                ("nil", TokenType::Nil),
                ("or", TokenType::Or),
                ("print", TokenType::Print),
                ("return", TokenType::Return),
                ("super", TokenType::Super),
                ("this", TokenType::This),
                ("true", TokenType::True),
                ("var", TokenType::Var),
                ("while", TokenType::While),
            ]),
        }
    }

    /// Makes a token with the given kind, lexeme, and line number.
    fn make_token(&self, kind: TokenType) -> Token<'source> {
        let lexeme = &self.source[self.start..self.current];
        Token::make_token(kind, lexeme, self.line, self.start, self.current)
    }

    /// Skips whitespace characters in the source code.
    fn skip_whitespace(&mut self) {
        while self.current < self.source.len() {
            match self.peek() {
                b' ' | b'\r' | b'\t' => {
                    self.advance();
                }
                b'\n' => {
                    self.line += 1;
                    self.advance();
                }
                b'/' => {
                    if self.peek_next() == b'/' {
                        while self.peek() != b'\n' && self.current < self.source.len() {
                            self.advance();
                        }
                    } else {
                        self.advance();
                    }
                }
                _ => break,
            }
        }
    }

    /// Advances the current position in the source code.
    fn advance(&mut self) -> u8 {
        let c = self.peek();
        self.current += 1;
        c
    }

    fn match_char(&mut self, expected: u8) -> bool {
        if self.current >= self.source.len() {
            return false;
        }

        if self.source.as_bytes()[self.current] != expected {
            return false;
        }

        self.current += 1;
        true
    }

    /// Peeks at the next character in the source code.
    fn peek(&self) -> u8 {
        if self.current >= self.source.len() {
            return 0;
        }
        self.source.as_bytes()[self.current]
    }

    /// Peeks at the character after the next character in the source code.
    fn peek_next(&self) -> u8 {
        if self.current + 1 >= self.source.len() {
            return 0;
        }
        self.source.as_bytes()[self.current + 1]
    }

    fn string(&mut self) {
        while self.peek() != b'"' && self.peek() != 0 {
            self.advance();
        }
        if self.peek() == b'"' {
            self.advance();
        }
    }

    fn error(&mut self, message: &'source str) -> Token<'source> {
        Token {
            kind: TokenType::Error,
            lexeme: message,
            line: self.line,
            span: (self.start, self.current),
        }
    }

    fn number(&mut self) -> Token<'source> {
        while self.peek().is_ascii_digit() {
            self.advance();
        }
        if self.peek() == b'.' && self.peek_next().is_ascii_digit() {
            self.advance();
            while self.peek().is_ascii_digit() {
                self.advance();
            }
        }
        self.make_token(TokenType::Number)
    }

    fn identifier(&mut self) -> Token<'source> {
        while self.peek().is_ascii_alphanumeric() {
            self.advance();
        }
        // Try to make keyword
        if let Some(kw) = self.keywords.get(&self.source[self.start..self.current]) {
            self.make_token(*kw)
        } else {
            self.make_token(TokenType::Identifier)
        }
    }

    /// Scan the next token in sequence
    pub fn scan_token(&mut self) -> Token<'source> {
        self.skip_whitespace();
        self.start = self.current;
        if self.current >= self.source.len() {
            return self.make_token(TokenType::Eof);
        }

        match self.advance() {
            b'(' => self.make_token(TokenType::LeftParen),
            b')' => self.make_token(TokenType::RightParen),
            b'{' => self.make_token(TokenType::LeftBrace),
            b'}' => self.make_token(TokenType::RightBrace),
            b';' => self.make_token(TokenType::Semicolon),
            b',' => self.make_token(TokenType::Comma),
            b'.' => self.make_token(TokenType::Dot),
            b'-' => self.make_token(TokenType::Minus),
            b'+' => self.make_token(TokenType::Plus),
            b'/' => self.make_token(TokenType::Slash),
            b'*' => self.make_token(TokenType::Star),
            b'!' => {
                if self.match_char(b'=') {
                    self.make_token(TokenType::BangEqual)
                } else {
                    self.make_token(TokenType::Bang)
                }
            }
            b'=' => {
                if self.match_char(b'=') {
                    self.make_token(TokenType::EqualEqual)
                } else {
                    self.make_token(TokenType::Equal)
                }
            }
            b'<' => {
                if self.match_char(b'=') {
                    self.make_token(TokenType::LessEqual)
                } else {
                    self.make_token(TokenType::Less)
                }
            }
            b'>' => {
                if self.match_char(b'=') {
                    self.make_token(TokenType::GreaterEqual)
                } else {
                    self.make_token(TokenType::Greater)
                }
            }
            b'"' => {
                self.string();
                self.make_token(TokenType::String)
            }
            c if c.is_ascii_digit() => self.number(),
            c if c.is_ascii_alphabetic() || c == b'_' => self.identifier(), // Also handles keywords
            _ => self.error("Unexpected character."),
        }
    }
}
